File: app/src/main/java/com/zsemberidaniel/egerbuszuj/fragments/ChooseStopFragment.java
package com.zsemberidaniel.egerbuszuj.fragments;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.zsemberidaniel.egerbuszuj.R;
import com.zsemberidaniel.egerbuszuj.adapters.ChooseStopAdapter;
import com.zsemberidaniel.egerbuszuj.realm.objects.Stop;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.TreeSet;

import eu.davidea.flexibleadapter.FlexibleAdapter;
import io.realm.Realm;
import io.realm.RealmResults;

/**
 * Created by zsemberi.daniel on 2017. 05. 12..
 */

public class ChooseStopFragment extends Fragment {

    private RecyclerView recyclerView;
    private RecyclerView.LayoutManager layoutManager;

    private FlexibleAdapter<ChooseStopAdapter.ChooseStopItem> allStopsAdapter;
    private TreeSet<ChooseStopAdapter.ChooseStopItem> allStopItems;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.choose_stop, container, false);

        recyclerView = (RecyclerView) view.findViewById(R.id.chooseStopRecyclerView);

        // improve performance because the content does not change the layout size of the RecyclerView
        recyclerView.setHasFixedSize(true);

        // use a linear layout manager
        layoutManager = new LinearLayoutManager(view.getContext());
        recyclerView.setLayoutManager(layoutManager);

        // specify an adapter with all of the stops
        RealmResults<Stop> allStops = Realm.getDefaultInstance().where(Stop.class).findAll();

        // Get letter headers
        HashMap<Character, ChooseStopAdapter.ChooseStopHeader> headers =
                ChooseStopAdapter.getNewHeaders(allStops);

        // Get the items in the adapter
        allStopItems = ChooseStopAdapter.convertToChooseStopItems(allStops, headers);

        allStopsAdapter = new ChooseStopAdapter(new ArrayList<>(allStopItems));
        allStopsAdapter.setDisplayHeadersAtStartUp(true);
        allStopsAdapter.setStickyHeaders(true);

        allStopsAdapter.addListener(new FlexibleAdapter.OnItemClickListener() {
            @Override
            public boolean onItemClick(int position) {
                ChooseStopAdapter.ChooseStopItem item = allStopsAdapter.getItem(position);


                return false;
            }
        });

        recyclerView.setAdapter(allStopsAdapter);

        return view;
    }

    public void updateStopFilter(String newFilter) {
        allStopsAdapter.setSearchText(newFilter);

        // Disable headers if the user is searching because they are not really needed
        // And if enabled they get doubled for some reason
        if (!newFilter.equals(""))
            allStopsAdapter.hideAllHeaders();
        else
            allStopsAdapter.showAllHeaders();

        allStopsAdapter.filterItems(new ArrayList<>(allStopItems));
    }
}

File: app/src/main/java/com/zsemberidaniel/egerbuszuj/fragments/ChooseStopFragment.kt
package com.zsemberidaniel.egerbuszuj.fragments

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.zsemberidaniel.egerbuszuj.R
import com.zsemberidaniel.egerbuszuj.adapters.ChooseStopAdapter
import com.zsemberidaniel.egerbuszuj.interfaces.views.IChooseStopView
import com.zsemberidaniel.egerbuszuj.presenters.ChooseStopPresenter

import eu.davidea.flexibleadapter.FlexibleAdapter
import java.util.*

/**
 * Created by zsemberi.daniel on 2017. 05. 12..
 */

class ChooseStopFragment : Fragment(), IChooseStopView {

    private lateinit var recyclerView: RecyclerView
    private lateinit var layoutManager: RecyclerView.LayoutManager

    private lateinit var allStopsAdapter: ChooseStopAdapter

    private lateinit var presenter: ChooseStopPresenter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.choose_stop, container, false)

        recyclerView = view.findViewById(R.id.chooseStopRecyclerView) as RecyclerView

        // improve performance because the content does not change the layout size of the RecyclerView
        recyclerView.setHasFixedSize(true)

        // use a linear layout manager
        layoutManager = LinearLayoutManager(view.context)
        recyclerView.layoutManager = layoutManager

        allStopsAdapter = ChooseStopAdapter(ArrayList<ChooseStopAdapter.ChooseStopItem>())
        allStopsAdapter.setDisplayHeadersAtStartUp(true)
        allStopsAdapter.setStickyHeaders(true)

        presenter = ChooseStopPresenter(this, allStopsAdapter)
        presenter.init()

        allStopsAdapter.addListener(FlexibleAdapter.OnItemClickListener { position ->
            presenter.stopClicked(context, allStopsAdapter.getItem(position))
            true
        })

        recyclerView.adapter = allStopsAdapter

        return view
    }

    override fun onResume() {
        super.onResume()
        allStopsAdapter.clear()
        allStopsAdapter.addItems(0, presenter.getAdapterItemsCopy())
    }

    override fun updateStopFilter(newFilter: String) {
        allStopsAdapter.searchText = newFilter

        // we need this because otherwise the items just disappear if the filter is "" for some reason
        if (newFilter == "") {
            allStopsAdapter.clear()
            allStopsAdapter.addItems(0, presenter.getAdapterItemsCopy())
            allStopsAdapter.showAllHeaders()
            return
        }
        allStopsAdapter.filterItems(presenter.getAdapterItemsCopy())

        // Disable headers if the user is searching because they are not really needed
        // And if enabled they get doubled for some reason
        if (newFilter != "")
            allStopsAdapter.hideAllHeaders()
    }
}
